### ISSUE TITLE
Connection re-use doesn't work due to modification of the options hash

### ISSUE THREAD

**Issue by Sharpie** [inspec/train#177]
During calls to `#connection`, the sub-classes of Transport will return an established connection if the option hash stored at time of creation matches the current options. However, the subclasses of BaseConnection mutate this hash by deleting items from it --- this prevents connection reuse from occurring.

**Comment by clintoncwolfe** [comment]
Hi! This issue has sat idle for a while, so I wanted to check two things:
1. Is this still relevant?
2. If so, could you describe your use case in more detail?

**Comment by Sharpie** [comment]
I think it is still relevant as the troublesome code is still present in some of the transports.

I recall the use case was simply "reusing connections". It looks like the hash on the right side of the equality comparison is generated from a stable function call, but the contents of the `@connection_options` instance variable are mutated by calls to `Connection.new`. So equality is just never possible and a new connection gets created for each command execution.

For example, in WinRM the options hash is created from `connection_options`, which is a stable function:

https://github.com/inspec/train/blob/v1.4.31/lib/train/transports/winrm.rb#L67

This hash is passed to `create_new_connection` which stores a reference to the hash in `@connection_options` and then passes it down to `Connection.new`:

https://github.com/inspec/train/blob/v1.4.31/lib/train/transports/winrm.rb#L130-L131

Initialization of the WinRM Connection mutates the option hash:

https://github.com/inspec/train/blob/v1.4.31/lib/train/transports/winrm_connection.rb#L34-L38

This mutation is reflected in `@options_hash` which means you pay the price of setting up a new WinRM connection for every command executed when the existing connection should just be re-used.

This was a year's worth of code ago for me and I'm no longer using Train in the project where I hit it, so I can't verify directly that the latest version is still affected.

**Comment by Sharpie** [comment]
The BaseConnection class also deletes the logger from the options array:

https://github.com/inspec/train/blob/master/lib/train/plugins/base_connection.rb#L24

So, this likely affects any transport that decides connection re-use using logic similar to:

https://github.com/inspec/train/blob/v1.4.31/lib/train/transports/winrm.rb#L69-L73

**Comment by kekaichinose** [comment]
Thank you for tracking down this issue. As with many large open source projects, we have a backlog of bugs and issues of varying user impact. Some issues are perpetually subject to lower priority than others in our backlog. We have reviewed the issue, and we have determined this a very low priority when considered in the context of the entire backlog. We have elected to close it as "Declined" due to competing priorities. We very much appreciate you bringing this to our attention.

Thank you!
__Chef InSpec Team__
